src/fileformats/ocd_types.cpp

```diff
@@ -100,6 +102,114 @@ namespace Ocd
 	}
 	
 	
+	// Cf. http://giflib.sourceforge.net/whatsinagif/lzw_image_data.html
+	IconV9 IconV8::uncompress() const
+	{
+		IconV9 result = {};
+		constexpr auto last = std::extent<decltype(result.bits)>::value;
+		auto pos = 0u;  // index in result.bits
+		
+		// Entries simply refer to the sequence which already exists in the output.
+		struct DictEntry {
+			const quint8* start;
+			unsigned int count;
+		};
+		
+		// Input size is 256 bytes and even less 9-bit code words. 254 entries is enough.
+		std::vector<DictEntry> dict;
+		dict.reserve(254);
+		
+		// Initialization handles first code word.
+		result.bits[pos++] = bits[16];
+		quint16 code;                              // current code word
+		auto carry = quint16(bits[17] >> 1);       // bits carried to next code word
+		auto carry_bits = quint8(7);               // number of bits carried to next code word
+		DictEntry prev_code = { result.bits, 1 };  // expansion of previous code word
+		// Compressed data starts at offset 16 of 264 bytes, and
+		// initialization consumed first two bytes.
+		for (int i = 18; i < 264; ++i)
+		{
+			// Get 9-bit code words from input bytes by filling the code word
+			// code word from the lowest to the highest bit.
+			auto input = bits[i];
+			if (!carry_bits)
+			{
+				carry = input;
+				carry_bits = 8;
+				continue;
+			}
+			Q_ASSERT(carry_bits <= 8);
+			code = carry + quint16((input & (2u << (8 - carry_bits)) - 1) << carry_bits);
+			carry = input >> (9 - carry_bits);
+			--carry_bits;
+			
```

Comment:
How about encapsulating the bit handling in a local class so that these details do not leak into the inner loop?

E.g.:
```
...
		dict.reserve(254);
		
		// bit handling code
		constexpr auto bits_in_code = 9;
		class {
			quint32 data {0};
			unsigned bits_left {0};
		public:
			// returns true when more bits are needed
			bool refill(unsigned b, unsigned min_watermark) {
				data |= (b << bits_left);
				bits_left += 8;
				return bits_left < min_watermark;
			}
			quint32 fetch(unsigned n_bits)
			{
				auto ret = data & ((1 << n_bits) - 1);
				data >>= n_bits;
				bits_left -= n_bits;
				return ret;
			}
		} accumulator;

		// handle the first code (haven't found a way how to fold it into the loop)
		accumulator.refill(bits[16], bits_in_code);
		accumulator.refill(bits[17], bits_in_code);
		result.bits[pos++] = quint8(accumulator.fetch(bits_in_code)); // yes, we don't check that first code is < 0x100
		DictEntry prev_code { result.bits, 1 };

		// Compressed data starts at offset 16 of 264 bytes, first two bytes
		// have already been handled
		int i = 18;
		while (i < 264)
		{
			// take bits from compressed stream and feed them into accumulator
			// (FIXME?) we are relaxed at this point and don't guard against overruns in the loop
			while (Q_LIKELY(accumulator.refill(bits[i++], bits_in_code)));
			auto code = accumulator.fetch(bits_in_code);

			// Handle code word
			if (code == 0x101)
....
```